Add pagination to followers and following page in dashboard (#258)

johncurcio opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

<!--- For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments -->

## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Bug Fix

## Description

Add infinite scroll to followers and following pages as described in #258 . In order to add pagination to the ``following`` page it was necessary to break it into 4 different subpages: tags, podcasts, organization and users, and each page has its own infinite scroll.

This is a much bigger change than I initially anticipated (maybe it should've been split into couple of PRs), so it probably needs some discussion and validation, especially regarding the UI changes to the following tab. 

What is done in this PR:

- [x] Add infinite scrolling to the followers page
- [x] Separate following page in 4 subsections and add infinite scrolling to to each subsection
- [x] Add acceptance tests to infinite scrolling functionality and unit tests to API and controller changes
- [x] Refactor initScrolling.js.erb  a bit.

## Related Tickets & Documents

Closes #258 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

![subpages](https://user-images.githubusercontent.com/9089551/66668139-e2d67d80-ec2a-11e9-8da4-f53e5631bf01.png)

![subpages-mobile](https://user-images.githubusercontent.com/9089551/66668142-e538d780-ec2a-11e9-9946-7187939a7cbb.png)

![following](https://user-images.githubusercontent.com/9089551/66667678-f6cdaf80-ec29-11e9-9fb7-874b52a7dbf3.gif)

![followers](https://user-images.githubusercontent.com/9089551/66667680-f9300980-ec29-11e9-8395-43acafbeebed.gif)

## Added to documentation?

- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

rhymes commented on spec/system/dashboards/user_scrolls_down_dashboard_follows_spec.rb:
```diff
@@ -0,0 +1,85 @@
+  let(:default_per_page) { 81 }
+  let(:users) { create_list(:user, default_per_page) }
+  let(:tags) { create_list(:tag, default_per_page) }
+  let(:organizations) { create_list(:organization, default_per_page) }
```
this is another reason to have a parameterized `per_page`, so you can create 5 or 10 items, instead 81 per each type, it will considerably speed up the test.

Also, see if you can use `let_it_be` instead of `let`, you can do so if you're not modifying the object in the tests

johncurcio replied:
I get a weird error when I try to use ``let_it_be``, which only happens when I run the entire test suite. I'm not sure what's causing it, because it does not fail when I run my test alone. I'll use ``let`` for now and change it to ``let_it_be`` if I figure out what's causing this behavior.

```ruby
Failure/Error: let_it_be(:tags, reload: true) { create_list(:tag, default_per_page) }
     
     Algolia::AlgoliaProtocolError:
       Cannot reach any host: 

  ================================================================================

       An HTTP request has been made that VCR does not know how to handle:
         PUT https://zox5q9ty1r.algolia.net/1/indexes/Tag_test/79174
     
       There is currently no cassette in use. There are a few ways
       you can configure VCR to handle this request:
     
         * If you're surprised VCR is raising this error
           and want insight about how VCR attempted to handle the request,
           you can use the debug_logger configuration option to log more details [1].
         * If you want VCR to record this request and play it back during future test
           runs, you should wrap your test (or this portion of your test) in a
           `VCR.use_cassette` block [2].
         * If you only want VCR to handle requests made while a cassette is in use,
           configure `allow_http_connections_when_no_cassette = true`. VCR will
           ignore this request since it is made when there is no cassette [3].
         * If you want VCR to ignore this request (and others like it), you can
           set an `ignore_request` callback [4].
     
       [1] https://www.relishapp.com/vcr/vcr/v/5-0-0/docs/configuration/debug-logging
       [2] https://www.relishapp.com/vcr/vcr/v/5-0-0/docs/getting-started
       [3] https://www.relishapp.com/vcr/vcr/v/5-0-0/docs/configuration/allow-http-connections-when-no-cassette
       [4] https://www.relishapp.com/vcr/vcr/v/5-0-0/docs/configuration/ignore-request
```

rhymes replied:
It's possible that it's a precedence problem, `let_it_be` is basically a before all: https://test-prof.evilmartians.io/#/let_it_be

Maybe you can get away with the `let!` ?

johncurcio commented on app/views/api/v0/follows/followers.json.jbuilder:
```diff
@@ -0,0 +1,8 @@
+  json.name                    follow.follower.name
+  json.path                    follow.follower.path
+  json.username                follow.follower.username
+  json.profile_image           ProfileImage.new(follow.follower).get(60)
```
@rhymes, I noticed while writing the partial that on the dashboard organization shows the name twice:

![org](https://user-images.githubusercontent.com/9089551/67128847-4c2e3180-f1d3-11e9-8880-03c3cb67eddd.png)

The username is ``azure``, so I was expecting it to show ``@azure`` but got ``@Microsoft Azure`` instead. Shouldn't it be ``@organization.username`` instead of ``@organization.name``? Or is there a reason why the name is used?

rhymes replied:
looks like a bug to me :)

johncurcio commented on app/assets/javascripts/initializers/initScrolling.js.erb:
```diff
@@ -6,15 +6,15 @@ function checkIfNearBottomOfPage() {
   var elCheck = document.getElementById("index-container");
```
Sure. I'm going to resolve the conflicts I got on this file, and then I'm going to see what I can refactor to remove the linting errors.

johncurcio replied:
@rhymes, I fixed the errors and warnings in this file (they're here https://gist.github.com/johncurcio/8de6a546ebc0a1cd6c4ed0a1fb67bbfc), however I feel like a couple of linting rules made the file harder to read or are unnecessary:

1. **no-use-before-define**: ``initScrolling`` is kind of the main function in this file, and I think it should be at the top, but because it indirectly calls ``fetchNextPageIfNearBottom`` which calls most functions in the file, ``initScrolling`` has to be placed at the bottom. I think that makes it harder to understand where to start reading the functions.

2. **no-console**: is it a problem to have the console logging an error?

I think we could ignore a few of these linting rules, unless there's a better way to do this that I'm not seeing. What do you think?

rhymes replied:
I understand your point but I think it's fine, it follows a pattern (maybe not an ideal one) but one that's consistent in all vanilla JS file we have. When we'll be able to move that code over to webpack and use ES6 modules it'll be clearer what's the entry point :)

Regarding the console error I think it's another thing we might have to live with right now, unless we actually implement a better error handling there, it's probably better to log the error instead of hiding 👀

rhymes replied:
I also think we should add `use strict` to this file

rhymes commented:
Hi @johncurcio, sorry again if we're moving slowly with this. Could you take a look at the conflict? I'll review it again afterwards so that we can take this to the finish line :)

johncurcio commented:
Hey @rhymes, don't worry about it. I fixed the conflict and added a couple of tests. I'm going to rebase my branch as soon as I finish running all tests.

rhymes commented on app/controllers/api/v0/followers_controller.rb:
```diff
@@ -0,0 +1,33 @@
+  module V0
+    class FollowersController < ApiController
+      before_action :authenticate_with_api_key_or_current_user!, only: %i[organizations users]
+      before_action -> { limit_per_page(default: 80, max: 1000) }, only: %i[organizations users]
```
I don't think you need the only clause, since there are only those two methods :)

rhymes commented on app/controllers/api/v0/followings_controller.rb:
```diff
@@ -0,0 +1,51 @@
+  module V0
+    class FollowingsController < ApiController
+      before_action :authenticate_with_api_key_or_current_user!, only: %i[tags users organizations podcasts]
+      before_action -> { limit_per_page(default: 80, max: 1000) }, only: %i[tags users organizations podcasts]
```
I don't think you need the only clause, since they apply to all methods :)

rhymes commented on app/controllers/dashboards_controller.rb:
```diff
@@ -1,7 +1,8 @@
   before_action :authenticate_user!
-  before_action :fetch_and_authorize_user, only: %i[show following followers]
+  before_action :fetch_and_authorize_user, only: %i[show following_tags following_users following_organizations following_podcasts followers]
+  before_action -> { limit_per_page(default: 80, max: 1000) }, only: %i[following_tags following_users following_organizations following_podcasts followers]
```
Probably easier using `except:` here :)
